exec/java-exec/src/main/java/org/apache/drill/exec/rpc/control/ControlClient.java

```diff
@@ -89,14 +90,48 @@ public MessageLite getResponseDefaultInstance(int rpcType) throws RpcException {
   }
 
   @Override
-  protected Response handle(ControlConnection connection, int rpcType, ByteBuf pBody, ByteBuf dBody) throws RpcException {
-    return handler.handle(connection, rpcType, pBody, dBody);
+  protected void handle(ControlConnection connection, int rpcType, ByteBuf pBody, ByteBuf dBody,
+                        ResponseSender sender) throws RpcException {
+    connection.getCurrentHandler().handle(connection, rpcType, pBody, dBody, sender);
   }
 
   @Override
   protected void validateHandshake(BitControlHandshake handshake) throws RpcException {
     if (handshake.getRpcVersion() != ControlRpcConfig.RPC_VERSION) {
-      throw new RpcException(String.format("Invalid rpc version.  Expected %d, actual %d.", handshake.getRpcVersion(), ControlRpcConfig.RPC_VERSION));
+      throw new RpcException(String.format("Invalid rpc version.  Expected %d, actual %d.",
+          handshake.getRpcVersion(), ControlRpcConfig.RPC_VERSION));
+    }
+
+    if (handshake.getAuthenticationMechanismsCount() != 0) { // remote requires authentication
+      if (config.getAuthProvider() == null) {
+        throw new RpcException(String.format("Drillbit (%s) requires auth, but auth is not configured.",
+            remoteEndpoint.getAddress()));
+      }
+      if (!handshake.getAuthenticationMechanismsList().contains(config.getAuthMechanismToUse())) {
+        throw new RpcException(String.format("Drillbit (%s) does not support %s", remoteEndpoint.getAddress(),
+            config.getAuthMechanismToUse()));
+      }
+
+      final SaslClient saslClient;
+      try {
+        saslClient = config.getAuthProvider()
```

Comment:
why the config doesn't return directly the factory? it would make the code a bit simpler I guess (and probably no need then to expose the mechanism used from outside the config)